odoo/addons/base/models/qweb.py

```diff
@@ -251,115 +109,120 @@ def _render(self, template, values=None, **options):
-            * ``profile`` (float) profile the rendering (use astor lib) (filter
-              profile line with time ms >= profile)
+            * ``load`` (function) overrides the load method (returns: (template, ref))
+            * ``profile`` (boolean) profile the rendering
         """
-        values = values or {}
-        body = []
-        self.compile(template, options)(self, body.append, values)
-        joined = u''.join(body)
-        if not values.get('__keep_empty_lines'):
-            joined = QWeb._empty_line.sub('\n', joined.strip())
-        return joined.encode('utf8')
+        render_template = self._compile(template, options)
+        return render_template(self, values or {})
 
-    def compile(self, template, options):
-        """ Compile the given template into a rendering function::
+    def _compile(self, template, options):
+        """ Compile the given template into a rendering function (generator)::
 
-            render(qweb, append, values)
+            render(qweb, values)
 
-        where ``qweb`` is a QWeb instance, ``append`` is a unary function to
-        collect strings into a result, and ``values`` are the values to render.
+        where ``qweb`` is a QWeb instance and ``values`` are the values to render.
         """
         if options is None:
             options = {}
 
+        element, document, ref = self._get_template(template, options)
+        if not ref:
+            ref = element.get('t-name', str(document))
+
+        options['ref'] = ref
+        if 'profile' in options:
+            options['document'] = document if isinstance(document, str) else str(document, 'utf-8')
+
         _options = dict(options)
         options = frozendict(options)
 
-        element, document = self.get_template(template, options)
-        name = element.get('t-name', 'unknown')
-
         _options['template'] = template
-        _options['ast_calls'] = []
         _options['root'] = element.getroottree()
         _options['last_path_node'] = None
         if not options.get('nsmap'):
             _options['nsmap'] = {}
 
-        # generate ast
+        def manage_exception(e, path, code=None):
+            if not options.get('dev_mode'):
+                code = None
+            element = self._get_template(template, options)[0]
+            html = None
+            if path and ':' not in path:
+                nodes = element.getroottree().xpath(path)
+                if nodes:
+                    node = nodes[0]
+                    while len(node) > 0: # remove children for the error log
+                        node.remove(node[0])
+                    html = etree.tostring(node, encoding='unicode')
+            return QWebException("Error when compiling template code", e, path, html, template, code=code)
+
+        if not self._builtins_names:
+            globals_dict = self._prepare_globals({}, {})
+            self._builtins_names.extend((globals_dict['__builtins__']))
```

Comment:
That looks suspicious and weird.

Reply:
To provide globals (which are secure from ir_qweb) used by compiled code and by compiled directives. Otherwise I can name everything (values ['item']) and add the globals in the dictionary of values

Reply:
@odony ping

Reply:
OK yeah I'd missed that there's one level of deref' in there, sorry about that.